torch_xla/experimental/scan.py

```diff
@@ -82,14 +132,514 @@ def scan(fn, init, xs):
   if xs_length is None:
     raise ValueError(f"`xs` {xs} is an empty PyTree.")
 
-  carry = init
-  ys = []
+  forward, backward = value_and_grad_partitioned(
+      fn, init, xs, partition_fn=partition_fn)
+  carry, ys = Scan.apply(forward, backward, init, xs)  # type: ignore
+  return carry, ys
+
+
+def value_and_grad_partitioned(
+    fn: Callable[[Carry, X], tuple[Carry, Y]],
+    init: Carry,
+    xs: X,
+    partition_fn=default_partition) -> tuple[Callable, Callable]:
+  """
+  Given a user `fn` to be scanned over the major axis of the input `xs` PyTree
+  and an initial carry object `init`, symbolically traces `fn` and returns two
+  functions, `forward` and `backward`, where `forward` packs intermediate
```

Comment:
`pack` do you mean `torch.stack` or something else?

Reply:
It just puts them in a large tuple.

Whereas the original `fn` outputs `(carry, y)`,
the transformed `forward` outputs `(carry, (y, activations))`

When we scan the transformed `forward` over the input, we'll get back not just the output, but also the stacked activations that could be stored in `ctx` and later used in the backward pass.